lgc/builder/ImageBuilder.cpp

```diff
@@ -1831,3 +1836,36 @@ Value *ImageBuilder::handleFragCoordViewIndex(Value *coord, unsigned flags, unsi
 
   return coord;
 }
+
+// =====================================================================================================================
+// Handle the uniform index by track back through a descriptor load and create a readfirstlane for it
+//
+// @param imageSampleInst : The instruction to get the image and sampler descriptors
+// @param operandIdx :
+// @param instName : Name to give instruction(s)
+void ImageBuilder::handleUniformIndex(Instruction *imageSampleInst, ArrayRef<unsigned> operandIdxs,
+                                      const Twine &instName) {
+  // For each uniform input, try and trace back through a descriptor load to find the uniform index used in it.
+  // If that fails, we won't pass the index to readfirstlane.
+  SmallVector<Value *, 2> uniformIndices;
+  findDescriptorIndices(imageSampleInst, operandIdxs, uniformIndices);
+
+  // Save Builder's insert point, and set it to insert new code just before each mul instruction.
+  auto savedInsertPoint = saveIP();
+
+  for (Value *uniformIdx : uniformIndices) {
+    // uniformIdx is a mul instruction that mulitplies an index from a descriptor load by the corresponding stride
+    if (auto mulInst = dyn_cast<BinaryOperator>(uniformIdx)) {
+      if (mulInst->getOpcode() == Instruction::Mul) {
+        SetInsertPoint(mulInst);
+        assert(!isa<Constant>(mulInst->getOperand(0)));
+        Value *index = CreateMul(mulInst->getOperand(0), mulInst->getOperand(1));
+        index = CreateIntrinsic(Intrinsic::amdgcn_readfirstlane, {}, index);
+        mulInst->replaceAllUsesWith(index);
+      }
+    }
+  }
```

Comment:
This is fragile because it may no longer be correct if e.g. CSE or GVN was run before this pass.

Reply:
See #840 for an elaboration on this.